### DISCUSSION TITLE
About WolfVision's future

### DISCUSSION THREAD

**Discussion by iChizer0** [wildwolf-team/WolfVision#19] [Announcements]
## Meaning

As one of the maintainers of this project, I didn't really understand the point of creating this project from the beginning of WolfVision. First of all, it should be noted that this project is actually a modification of [rm-2021-Code-Framework](https://github.com/rcxxx/rm-2021-Code-Framework), the new project retains its predecessor's project structure and design ideas, based on subdivision and refined, it has the prototype of the current project. The name is composed by our team name "WildWolf" and the main function it implements: "Vision", but it does not stand for something like "Wolf's vision (sight)", in fact wolfs vision (sight) is not sharp.

作为这个项目的维护者之一，从 WolfVision 诞生之初，我其实并不能十分理解创造这个项目的意义。首先需要说明一下，这个项目其实是在 [rm-2021-Code-Framework](https://github.com/rcxxx/rm-2021-Code-Framework) 上修改而来的，新的项目保留了其前身的项目结构和设计思路，在这个基础上加以细分和完善，就有了当前项目的雏形，名字的组成是由我们的队名 WildWolf 和它实现的主要功能 Vision 组成的，它并不是代表像 “狼的视觉”，事实上狼的视觉并不敏锐。

During these days of maintenance, I would like to share some personal thoughts about the meaning of creating this project.

- To provide new members of the Visual Algorithms Group with exposure to more open source projects and to familiarize them with the process of using the tools for maintaining these open source projects, and to learn how to maintain open source projects and collaborate with other contributors

- Provide a well-structured, easy-to-maintain, modular framework for existing members of the Visual Algorithms Group to improve code quality, achieve better competition results

- Attract good contributors to make improvements to our code (small probability)

在这几天的维护过程中，我想分享一下个人关于创造这个项目意义的想法：

- 为视觉算法组新成员提供接触更多开源项目的机会，并让他们熟悉维护这些开源项目的工具的使用流程，了解如何维护开源项目以及与其它贡献者合作

- 为视觉算法组现有成员提供一个结构清晰、易于维护、模块化的框架，提高代码质量，取得更好的比赛成绩

- 吸引优秀的贡献者，为我们的代码做出改进（小概率）

## Problems

Obviously that we created this project with a clear purpose. But in achieving that purpose, we encountered many fundamental problems that made it difficult to grow the project, which is why I'm here to discuss the future of the project.

显然，我们创造这个项目是有明确的目的性的。但是在实现这个目的性时，我们遇到了许许多多根本性的问题，以至于会导致这个项目难以发展下去，这也是我为什么要在这里讨论这个项目的未来。

First, the code architecture of this project is still problematic by design. We repeatedly emphasize abstraction and encapsulation, which is actually reflected in the existing code: abstract for abstraction and encapsulate for encapsulation. My suggestion is to merge `module/armor` and `module/buff` into `module/target`, with target as a separate module containing two submodules armor and buff, and the abstract class abstract, with the following directory structure tree is here:

首先，这个项目的代码架构在设计上仍然是存在问题的。我们一再强调抽象、封装，实际上在现有代码中体现出来的是：为了抽象而抽象，为了封装而封装。我的建议是，将 `module/armor` 和 `module/buff`，合并到 `module/target` 中，以 target 作为一个独立模块，其包含两个子模块 armor 和 buff，以及抽象类 abstract，目录结构树：

```
module
└── target
    ├── abstract
    │ └── light.hpp
    ├─ armor
    │ ├─ armor.cpp
    │ └── armor.hpp
    ├─ buff
    │ ├─ buff.cpp
    │ └── buff.hpp
    ├── target.cpp
    └── target.hpp
```

The main reason for abstracting the light bar is which has good reuse value for armor and buff in identifying the hit targe process (color, fitting method, fitting shape), the fitting configuration of the light bar could be defined as `Light_Config` and the gain parameters could be defined as `Light_Data`, The new requirement of the data structure of the project here is to establish a unified exchangeable data type and to limit the code execution process by limiting the data types.

抽象 light 灯条的主要原因是它在识别击打 targe 过程中，对于 armor 和 buff 有良好的复用价值（颜色、拟合方式、拟合形状），灯条的拟合配置可定义为 `Light_Config`，获得的参数可定义为 `Light_Data`，这里对项目的数据结构提出了新的要求，即建立统一的可交换的数据类型，通过限定数据类型来限制代码执行流程。

Second, the current code uses only the most basic C++ features, and the problem is focused on naming domains, class naming, and class member function naming. There are many function names that do not correspond to function functions or nonsensical nesting functions like `c(); // c() { a(); b(); }`, because of this we have to impose strict requirements on code quality, naming rules must follow *Google C++ Coding Style*, write meaningful, specific function names, and do not over-encapsulate.

其次，当前的代码仅仅使用到了最基本的 C++ 特性，问题集中在命名域、类命名和类成员函数命名上。出现函数名不对应函数功能或类似于 `c(); // c() { a(); b(); }` 的无意义套娃函数，正因如此我们要对代码质量提出严格的要求，命名法则一定要遵循 *Google C++ Coding Style*，写有意义、具体的函数名称，不做过度封装。

In addition, we should avoid direct coupling between each modules, as in `class A { B b(args_v); C c(args_v); }`, and to solve the problem of data exchange or functional stickiness, we introduce the concept of types.

此外，我们还应当避免各个模块之间的直接耦合，如 `class A { B b(args_v); C c(args_v); }`，为了解决数据交换或者功能上的粘连问题，我们引入 types 的概念。

In summary, we divide the new named domains into.

- `wv_device` device module
- `wv_module` function module
- `wv_types`  data structure module
- `wv_utils`  tool module and tools for data structure

综上所述，我们划分新的命名域：

- `wv_device` 设备模块
- `wv_module` 功能模块
- `wv_types`  数据结构模块
- `wv_utils`  工具模块和数据结构工具

The ideal code architecture design can be simply summarized as follows:

理想的代码架构设计可以简单概括为：

```cpp
#include "wolfvision.hpp"

int main() {
  wv::device::Camera     camera     (wolfvision::camera_config);
  wv::device::SerialPort serial_port(wolfvision::serial_config);

  wv::module::Car        car        (wolfvision::car_config);
  wv::module::Buff       buff       (wolfvision::buff_config);
  wv::module::RoI        roi        (wolfvision::roi_config);
  wv::module::Predict    predict    (wolfvision::predict_config);
  wv::module::Compensate compensate (wolfvision::compensate_config);

  wv::types::RoI_Data                           roi_data;
  wv::types::RingBuffer<wv::types::Camera_Data> camera_data_buffer(10);
  wv::types::RingBuffer<wv::types::Target_Data> target_data_buffer(10);

  while (1) {
    wv::types::Serial_Data serial_data { serial_port.getSerialData() };

    camera_data_buffer.emplace(camera.getCameraData(roi_data));

    wv::types::Target_Data target_data{
        wv::utils::TargetInit(
            buff.getTargetData(serial_data, camera_data_buffer, target_data_buffer),
            car .getTargetData(serial_data, camera_data_buffer, target_data_buffer))
    };

    target_data_buffer.emplace(target_data);

    wv::types::Target_Data target_data_fixed;

    target_data_fixed = predict   .getTargetData(serial_data, target_data_buffer);
    target_data_fixed = compensate.getTargetData(serial_data, target_data_buffer);
    roi_data          = wv::utils::Target2RoI(target_data_fixed);

    serial_port.sendSerialData(wv::utils::Target2Serial(target_data_fixed));
  }
}

```

**Comment by fqjun** [comment]
## Meaning

---

作为这个项目的维护者之一，也作为视觉算法组的早期成员之一，我是对该项目表示极大的赞赏。我先说明一下创建这个项目的意义，然后对上述讨论表达相应的见解并给予一定的建议。创建该项目的想法在2019年比赛结束后第一次提出，目的包含了以下几点：

- 为视觉算法组参赛队员建立一个属于自己的代码框架，能实现各个兵种都用同一份代码，方便各个队员进行协作修改以及各兵种的调试
- 为视觉算法组新成员提供良好的学习平台，方便后续迭代优化，而不是重复“造轮子”

## Proposals

---

针对提出的问题，我想表达我的一些想法：

首先，对于抽象、封装所出现的问题，个人觉得原因是在当前各个组员对抽象、封装的理解和使用方式不一，从而导致使用规范的问题频频出现，具体用法后面进行讲解，我的建议是：

- 统一一个具体的抽象、封装的标准
- `module`模块中仅包含功能调用和实现部分，不包括任何抽象对象 `target`、`light` 等, 独立功能包 `pnp`、`roi` 等,目录结构树：
  
```
module
├─ buff
│ ├─ buff.cpp
│ └── buff.hpp
├─ armor
│ ├─ armor.cpp
│ └── armor.hpp
└ engineer
     ├── engineer.cpp
     └── engineer.hpp

```

而 `types` 模块可以作为数据结构模块和抽象对象模块，用于存储各个功能所需的抽象对象，如 `armor` 所用到的 `light`，`buff`所用到的 `blade`、`center_r`等。因为该部分为各个队员需要开发的部分，每个人的实现任务的思路不一，不能有非常统一的实现做法，所以每个人所创建的对象可能都有所区别，只能根据过往经验总结和抽象出常见的属性从而进行封装，上述所提到的 `light` 就是一个典型的例子，该对象内会包含常见的属性，比如 `cv::RotateRect` , `angle` ,`index` ,`width` ,`height` 等,这些属性可以组成一个抽象类，便于重复利用。我的建议是，将常见的属性封装成一个抽象类，暂且命名为 `object` ，其他功能模块需要利用时，利用该抽象类来创建自己相应的对象目标，方便管理任务处理对象，目录结构树：

```
types
├── abstract
│ └── object.hpp
├─ blade
│ ├─ blade.cpp
│ └── blade.hpp
├─ light
│ ├─ light.cpp
│ └── light.hpp
├─ ore
│ ├─ ore.cpp
│ └── ore.hpp
└─ center_r
 ├─ center_r.cpp
 └── center_r.hpp

```

其次，我个人认为，各个模块之间直接耦合关系的评价和代码架构设计有关。结合上文所提出的新的架构，我简单概括一下我认为的代码架构设计：

![2021-07-13 03-44-10 的屏幕截图](https://user-images.githubusercontent.com/46771131/125346703-a0584c80-e38c-11eb-8dcd-4db76c21237d.png)


简单说明一下提出该代码架构的原因。以上位机的任务来作为切入点进行分析，上位机对于下位机的作用在接收下位机的数据并进行处理，最终返回数据给下位机，在执行任务期间，上位机对于下位机来讲可以充当一个运算平台，是一个黑盒子。在黑盒子中，我们会根据下位机所需的任务目标来选择执行的任务模式，其中包括自瞄、能量机关、矿石识别等，未来还会根据比赛任务进行拓展。而每个功能模块会根据任务需求使用已有的独立功能模块和工具模块来进行任务的处理，这时候开发队员只需要对已经成熟的独立功能模块和工具模块调用即可，如:需要 `pnp` 计算角度深度的则包含头文件即可;需要滤波处理的则包含 `filter`即可;尽量将心思花费在任务的逻辑实现上，所以就会有设备模块 `device`、数据结构模块 `types`、工具模块和数据结构工具 `utils`。但代价就是需要频繁修改 cmake 的配置文件，针对这一点我提出的建议是开发人员先在功能模块内进行创建及测试，待测试完善后才将部分数据结构模块和工具模块独立出来。

总结一下，尽管狼的视觉没那么敏锐，但我相信的是，如果大家能将精力投入在新功能的实现上而不是每年重复造“轮子”，那么我们队伍才会有进步的空间和时间。

**Reply by bug-hunter-cn** [reply]
支持支持(✪ω✪)

**Reply by iChizer0** [reply]
Detailed project is here: https://github.com/iChizer0/WolfVision/tree/dev

**Reply by iChizer0** [reply]
@robomaster-shl @roboxxltd @gcusms @JoyceKirkland
